lib/pos_leader_schedule.go

```diff
@@ -0,0 +1,106 @@
+package lib
+
+import (
+	"crypto/sha256"
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+)
+
+func (bav *UtxoView) GenerateLeaderSchedule() ([]*PKID, error) {
+	// Retrieve CurrentRandomSeedHash.
+	currentRandomSeedHash, err := bav.GetCurrentRandomSeedHash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error retrieving CurrentRandomSeedHash: ")
+	}
+
+	// Retrieve top, active validators ordered by stake.
+	validatorEntries, err := bav.GetTopActiveValidatorsByStake(int(bav.Params.LeaderScheduleMaxNumValidators))
```

Comment:
TODO: Need to make sure that this function returns the top validators in a deterministic order. If it has any non-determinism in it, then nodes will have different leader schedules.

Reply:
I have a test where I regenerate a leader schedule using the same CurrentRandomSeedHash 10x times. It generates the same leader schedule each time. With that test, I'm fairly confident in this algorithm's determinism.